[RDY] use guicursor to control cursor across UIs

teto opened the pull request:
Hi,

The PR is really messy at the moment but I send it anyway as I would like some feedback on the API before spending time on polishing. 
1st commit might work for shape but I changed some code in the second commit to take into account blink thus only insert mode should work (and for VTE) so it's useless to try it.

As I said earlier I am not sure if the API is correct so feel free to comment aggressively on the function prototypes you want (see  comments). I can adjust the rest.
.

teto commented on src/nvim/cursor_shape.c:
```diff
@@ -36,11 +38,61 @@ static cursorentry_T shape_table[SHAPE_IDX_COUNT] =
+    /*   case SHAPE_HOR: STRCPY(buf, "horizontal"); */
+    /*   default:; */
+    /* }; */
+    /* PUT(dic, "shape", STRING_OBJ(buf)); */
```
I believe that nvim prefers to be verbose over the msgpack-rpc, thus should I convert the shape values to strings ?

justinmk replied:
Yes. It's more a matter of not wanting to expose the internal order to the API, and then having to support those numbers forever more. It also is hopefully a better experience for API users.

teto commented on src/nvim/cursor_shape.c:
```diff
@@ -36,11 +38,61 @@ static cursorentry_T shape_table[SHAPE_IDX_COUNT] =
+    /* v */
+ 
+    /* PUT(dic, "name", STRING_OBJ(cur->name)); */
+    PUT(all, cur->name, DICTIONARY_OBJ(dic));
```
Should we pass this over an array of "cursorentry" or a dict of cursorentry with the mode as the key ?

teto commented on src/nvim/tui/tui.c:
```diff
@@ -69,6 +70,9 @@ typedef struct {
   bool busy;
+  /* use int instead ? */
+  // cursorentry_T cursor_shapes[SHAPE_IDX_COUNT]; /**!< is it fine to reuse cursorentry_T here ? */
+  int cursor_shapes[SHAPE_IDX_COUNT]; /**!< is it fine to reuse cursorentry_T here ? */
```
At first in tui_cursor_style_set, I converted the data back to cursorentry_T as in nvim core, hence the `cursorentry_T cursor_shapes[SHAPE_IDX_COUNT]` but then I thought it might better to generate directly the control sequence characters ?

justinmk replied:
Generate from where? All terminal sequence fiddling should live in tui.c.

teto replied:
As I see it, there are some members in cursorentry_T that don't make sense for the TUI (mshape/blinkoff ?). So I thought I could save a few bytes by generating directly in tui.c, upon cursor_shapes receptions the control sequence and discard the useless stuff. But if you intend to support blinkon/blinkoff in the TUI with timers, it makes sense to regenerate the cursorentry_T.

justinmk replied:
Oh. No, please don't do that. The TUI follows the same protocol as _all_ UIs. It cannot and must not receive different data than some other UI. Don't worry about saving bytes, especially because this is a rare event.

teto replied:
Of course, the conversion happened once the TUI received all data. Looks like it complexifies things instead of making them simpler so forget about it xD

teto commented on src/nvim/tui/tui.c:
```diff
@@ -434,6 +441,73 @@ static void tui_cursor_goto(UI *ui, int row, int col)
+    default: WLOG("Unknown shape value %d", shape); break;
+  };
+
+  // test for vte 
```
works only for vte . Just for test I consider it blinks if blinkon != 0

teto reviewed (commented):
These are the points I am most uncomfortable with.

Shougo commented:
This feature can change the cursor style from Vim script?

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -69,6 +70,9 @@ typedef struct {
   bool mouse_enabled;
   bool busy;
+  /* use int instead ? */
+  // cursorentry_T cursor_shapes[SHAPE_IDX_COUNT]; /**!< is it fine to reuse cursorentry_T here ? */
```
cursorentry_T can be used (like HlAttrs below). This is TUI's internal implementation, it never sees RPC. Think of it as its own "external UI".

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -682,6 +780,7 @@ static void tui_set_icon(UI *ui, char *icon)
 // to make a copy for the tui thread
 static void tui_event(UI *ui, char *name, Array args, bool *args_consumed)
 {
+  ILOG("TUI event");
```
This is only used for the externalized popupmenu. The name is a bit confusing.

teto commented on src/nvim/ui.h:
```diff
@@ -22,6 +23,7 @@ struct ui_t {
   void (*clear)(UI *ui);
   void (*eol_clear)(UI *ui);
   void (*cursor_goto)(UI *ui, int row, int col);
+  void (*cursor_style_set)(UI *ui, Dictionary args); // or a Dictionary ?
```
I forgot to ask this one: I don't think Dictionary will be available to other UIs so what should I use ?
(Dictionary defined in https://github.com/neovim/neovim/blob/9ef4be9aab6a9a16cf185ca17a2dabe87f8aa328/src/nvim/api/private/defs.h)

justinmk replied:
These parameters aren't directly mapped to RPC, there's some intermediate step. Notice how highlight_set has an HlAttrs parameter. So just use whatever is convenient internally.

justinmk replied:
The mapping is done in `api/ui.c` (all remote stuff is in `api/`).

teto replied:
I understood that correctly but your sugggestion in https://github.com/neovim/neovim/issues/2583#issuecomment-272988384 made me wonder.
I can only think of pushing each cursorentry via several calls of :
 
```
void (*cursor_style_set)(UI *ui,   int shape;                    /* one of the SHAPE_ defines */
  int mshape,                   /* one of the MSHAPE defines */
  int percentage,               /* percentage of cell for bar */
  long blinkwait,               /* blinking, wait time before blinking starts */
  long blinkon,                 /* blinking, on time */
  long blinkoff,                /* blinking, off time */
  int id,                       /* highlight group ID */
  int id_lm,                    /* highlight group ID for :lmap mode */
  char        *name,            /* mode name (fixed) */
char used_for, /* SHAPE_MOUSE and/or SHAPE_CURSOR */)
```

justinmk replied:
You can store the entire cursor style in a Dictionary (which may have nested Dictionaries).

teto replied:
But then it implies that UIs have a definition of the Dictionary type, and that they include nvim/api/private/defs.h, which doesn't seem to be currently the case (looking at you nvim-qt) ?

justinmk replied:
nvim-qt never sees any of this, it communicates via msgpack-RPC types. That's where the `api/ui.c` "intermediate step" comes into play.

Only in-process UIs would be affected by the choice of type here. Currently VimR is the only example of that. The `private/` namespace is for the theoretical future where we have a "pubic ABI": `api/*` excluding `api/private/*`. 

But in the particular case of `api/private/defs.h`, it's included by `api/buffer.h` and friends:

     #include "nvim/api/private/defs.h"

So Dictionary, Buffer, and the other API C types are (intentionally) exposed in the public ABI. The purpose of the `private/` namespace is to hide the implementation, not the types.

teto replied:
Of course silly me ! I had this in mind then lost track of things. Nothing holds me back now thanks.

teto commented:
@shougo the idea is to let the TUI cursor be set via "set guicursor". Providing a vimscript way to set the cursor should be trivial after that.

teto commented:
I cleaned up the code and it should  be working for Konsole and termite, It won't change the cursor for other terminals though as I was not sure how to do it for those terminals. The current code seems to use
```
  // Initialize the cursor shape.
  unibi_out(ui, data->unibi_ext.set_cursor_shape_block);
```
but I don't see how it's initialized :/

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -7,40 +7,93 @@
-
+///
+/// Handling of cursor and mouse pointer shapes in various modes.
+///
```
we don't have "border lines" in our docstring comments

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -7,40 +7,93 @@
-  {0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR},
+  /// The values will be filled in from the 'guicursor' and 'mouseshape'
+  /// defaults when Vim starts.
+  /// Adjust the SHAPE_IDX_ defines when making changes!
```
The 3-slash comment form is for doxygen. This isn't decorate a particular symbol so it should be the 2-slash form.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -434,6 +433,83 @@ static void tui_cursor_goto(UI *ui, int row, int col)
+      } else if (strcmp(shapeStr, "horizontal") == 0) {
+        r.shape = SHAPE_HOR;
+      } else {
+        WLOG("Unknown value %s", shapeStr);
```
Might want to make this user-visible, i.e. `EMSG2`

teto replied:
ok I'll use EMSG2(_(e_invarg2), shape_str)

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -434,6 +433,83 @@ static void tui_cursor_goto(UI *ui, int row, int col)
+      r.blinkon = value.data.integer;
+    } else if (strcmp(keyStr, "blinkoff") == 0) {
+      r.blinkoff = value.data.integer;
+    } else if (!strcmp(keyStr, "hl_id")) {
```
`== 0` like the others

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -434,6 +433,83 @@ static void tui_cursor_goto(UI *ui, int row, int col)
+{
+  TUIData *data = ui->data;
+  static char *modes[] = {
+    "normal",
```
Is this list TUI-specific, or should it live in cursor_shape.c?

teto replied:
It's just a way to map string to integer. Moving it to cursor_shape.c would be interesting only as a function int str_to_int_mode(char*) other UIs could call.

justinmk replied:
Sounds like that would be useful for e.g. VimR (which doesn't use a TUI, but is a fork of Nvim that uses ui_bridge.c to drive its GUI).

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -463,33 +539,74 @@ static void tui_mouse_off(UI *ui)
+
+    // odd => blink ?
+    ILOG("x1b[%d q (blinking= %d)", shape + (c.blinkon ==0), (c.blinkon != 0));
+    printf("\x1b[%d q", shape + (c.blinkon ==0));
```
this should go in the else{} above, otherwise it will reach konsole

teto replied:
I removed the ILOG statement but I think the printf is in the   `else if (!vte_version || atoi(vte_version) >= 3900) {` block. Maybe I shouldn't use printf but some unib_* call instead ?

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -7,40 +7,108 @@
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "vd", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "m", SHAPE_MOUSE},
-  {0, 0, 0,   0L,   0L,   0L, 0, 0, "ml", SHAPE_MOUSE},
-  {0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR},
```
I didn't realize that the short strings "n", "v", etc. could be used to identify these entries. Probably those should be used (over RPC) instead of creating the long-form names.

justinmk replied:
However, the long-form names would be useful as comments on each of these lines.

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -7,40 +7,108 @@
+/// @see cursor_mode_str2int
+static char *mode_names[] = {
+  "normal",
+  "visual",
```
These are already in `shape_table` so why is this extra table needed?

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -208,6 +276,22 @@ char_u *parse_shape_opt(int what)
+int cursor_mode_str2int(const char *mode)
+{
+  for (int current_mode = 0; mode_names[current_mode]; current_mode++) {
+    if (strcmp(mode_names[current_mode], mode) == 0) {
```
could use `shape_table[current_mode].name` instead.

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -7,40 +7,76 @@
- */
+/// Converts cursor_shapes into a Dictionary of dictionaries, of the form:
+/// "normal" : { "cursor_shape": ... }
+/// "insert" : { "cursor_shape": ... }
```
does the comment need to be updated?

justinmk commented on src/nvim/ui_bridge.c:
```diff
@@ -172,6 +173,17 @@ static void ui_bridge_cursor_goto(UI *b, int row, int col)
+  UI_CALL(b, cursor_styleset, 2, b, &args);
+}
+
+static void ui_bridge_cursor_styleset_event(void **argv)
```
This should "hug" `ui_bridge_cursor_styleset`, instead of `ui_bridge_cursor_goto_event`. IOW delete the newline above, and add one below :)

justinmk commented on src/nvim/ui.c:
```diff
@@ -373,6 +373,7 @@ void ui_cursor_style_set(void)
 {
   Dictionary style = cursor_shape_dict();
   UI_CALL(cursor_style_set, style);
+  api_free_dictionary(style);
```
Can't do this here. UI_CALL is asynchronous.

Shougo commented:
> @shougo the idea is to let the TUI cursor be set via "set guicursor". Providing a vimscript way to set the cursor should be trivial after that.

Nice.  Thanks.

fmoralesc commented:
It would be nice if this could change the cursor highlighting in the TUI too. As far as I can tell, it doesn't.

teto commented:
justinmk judged wiser to postpone cursor highlighting in a separate PR. I
agree it would be a neat feature.


2017-02-22 17:01 GMT+01:00 Felipe Morales <notifications@github.com>:

> It would be nice if this could change the cursor highlighting in the TUI
> too. As far as I can tell, it doesn't.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/6044#issuecomment-281712539>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA2FOtdmrmV0YDylvLeazWFJnMsIVuNaks5rfFvAgaJpZM4L0YGm>
> .
>

justinmk commented:
segfault in this PR reported at: https://github.com/neovim/neovim/issues/3688#issuecomment-281767725

teto commented:
To take into account @shougo's request (i.e., let vimscript be capable of setting cursor appearance regardless of the mode), I will need another ui call set_cursor_current_style(shape,blonkon,blinkoff, hl_groupid) ?

justinmk commented:
I guess so, but can't we wait on that and see if it's good enough to allow plugins to set 'guicursor'?

teto commented on src/nvim/ui_bridge.c:
```diff
@@ -181,12 +182,19 @@ static void ui_bridge_cursor_goto_event(void **argv)
+  Object copy = copy_object(DICTIONARY_OBJ(args));
+  Object *a = xmalloc(sizeof(Object));
+  *a = copy;
+  UI_CALL(b, cursor_styleset, 2, b, a);
```
am I right in doing a deep copy of the dictionary ? as the events can be deferred, I suppose I have to do it else we could send an freed memory in the event ?

justinmk replied:
who else is using `args`? Normally you should just pass a pointer and it is only freed when `ui_bridge_cursor_styleset_event` is reached.

justinmk replied:
Also the `*a = copy` won't work because data  (`copy`) was allocated on the stack. You would need to use `memcpy(a, &copy, sizeof(Object))`.

justinmk replied:
Based on `ui_bridge_put()` looks like that's the right approach.

teto replied:
The event can be deferred apparently in tui_scheduler so you have to make a copy else the original args is already freed when the event triggers.
*a = copy should work just fine in C, struct assignment is equivalent to a memcpy (?!): the deepcopy is done via the copy_object before.

justinmk commented on src/nvim/cursor_shape.h:
```diff
@@ -35,16 +35,17 @@
+  int id;                 //!< highlight group ID
+  int id_lm;              //!< highlight group ID for :lmap mode
+  char *name;             //!< mode short name
+  char used_for;          //!< SHAPE_MOUSE and/or SHAPE_CURSOR
```
what's the "!" for ? doxygen format is normally `///<` AFAIK

teto commented:
//!< is also valid doxygen. I've updated with ///<

2017-02-25 16:09 GMT+01:00 Justin M. Keyes <notifications@github.com>:

> *@justinmk* commented on this pull request.
> ------------------------------
>
> In src/nvim/cursor_shape.h
> <https://github.com/neovim/neovim/pull/6044#discussion_r103080649>:
>
> > -  long blinkoff;                /* blinking, off time */
> -  int id;                       /* highlight group ID */
> -  int id_lm;                    /* highlight group ID for :lmap mode */
> -  char        *name;            /* mode name (fixed) */
> -  char used_for;                /* SHAPE_MOUSE and/or SHAPE_CURSOR */
> +  char *full_name;        //!< mode full name
> +  int shape;              //!< cursor shape: one of the SHAPE_ defines
> +  int mshape;             //!< mouse shape: one of the MSHAPE defines
> +  int percentage;         //!< percentage of cell for bar
> +  long blinkwait;         //!< blinking, wait time before blinking starts
> +  long blinkon;           //!< blinking, on time
> +  long blinkoff;          //!< blinking, off time
> +  int id;                 //!< highlight group ID
> +  int id_lm;              //!< highlight group ID for :lmap mode
> +  char *name;             //!< mode short name
> +  char used_for;          //!< SHAPE_MOUSE and/or SHAPE_CURSOR
>
> what's the "!" for ? doxygen format is normally ///< AFAIK
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/neovim/neovim/pull/6044#pullrequestreview-23859803>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2FOqYFjypO8EjwRspDyw2cPF3MCDEeks5rgEQ6gaJpZM4L0YGm>
> .
>

teto commented:
I've been using this PR as my nvim for 2 days and I believe it's very close to being ready (I am afraid changing color might print some bytes on some terminals though as it's kinda a new feature: It worked on VTE + xterm). 

There is 1 (voluntary) linting error. 

I've had to add the last commit to prevent these errors:
```
	./test/functional/helpers.lua:136: in function 'run'
	./test/functional/ui/screen.lua:272: in function 'wait'
	./test/functional/ui/screen.lua:209: in function 'expect'
	/home/teto/neovim2/test/functional/ui/quickfix_spec.lua:45: in function </home/teto/neovim2/test/functional/ui/quickfix_spec.lua:10>

./test/functional/helpers.lua:136: ./test/functional/ui/screen.lua:316: attempt to call field '_on_event' (a nil value)
```
but you may want to replace it with some better placeholder.

teto commented:
I believe it's ready (I haven't touched the help though), I don't get the errors in https://travis-ci.org/neovim/neovim/jobs/207489328:
`profiling:/home/travis/build/neovim/neovim/build/src/nvim/CMakeFiles/nvim.dir/regexp.c.gcda:Merge mismatch for function 15`

frangio commented:
In urxvt the cursor does not blink by default. After closing neovim with this patch the cursor remains blinking.

frangio commented:
Changing the highlight group in `guicursor` doesn't do anything. Is that supposed to work?

Even if it's impossible to change the color of the cursor in the terminal, there's an intention to use this feature to make the cursor invisible in plugins such as incsearch.vim and EasyMotion (see #3688). A special case could be made for that.

For example, incsearch.vim [uses a `NONE` highlight group](https://github.com/haya14busa/incsearch.vim/blob/161c5b66542e767962ca5f6998a22e984f8d8a60/autoload/vital/_incsearch/Over/Commandline/Base.vim#L328) to hide the cursor. I don't think that one comes by default in Neovim. Maybe the `TermCursorNC` group could be used for the same purpose.

teto commented:
the highlight code is quite simple: https://github.com/neovim/neovim/pull/6044/files#diff-3ac88981a139ac0eee448a6bd19c7569R569. I believe NONE would be a special usecase as termcodes to hidecursor is sthg special. This is sthg I can look at

teto commented:
`nvim -u NORC -c ":echo hlID('NONE')` shows 0 so indeed it does not exist.
set guicursor does not seem to validate the highlight (it accepts undefined highlights) so we could say that when undefined the cursor becomes invisible but that might be very annoying for users in case of typo or if they use this to detect the loading of some plugin. It might be best to either introduce a new guicursor subsetting or abuse "hor" or "vert", ie. when set to 0, it hides the cursor.
@justinmk any opinion ?

justinmk commented:
Let's not worry about colors in this PR. BTW, I want to mention this note from Vim's `:help termcap-cursor-shape`:

      NOTE: When Vim exits the shape for Normal mode will remain.  The shape from
      before Vim started will not be restored.

teto committed:
Shape guicursor support for TUI

set one of blinkon/blinkoff to a positive value to enable blinking
cursor.

teto committed:
guicursor color support

For now only supports valid hex colors (does not check for the validity
 the hex color) when termguicolors is set, otherwise it won't
attempt to change the cursor color.

teto commented:
Compared to your last review, I've added some enums in cursor_shape.h because cursor vs editors modes start to be confusing; using an enum makes it clearer than int-based values. I rebased the work so that you can disable color support if you want but I would consider keeping it as it's very cool to have :) Ofc it's always possible to mess it up if you set a bad guibg value i.e.`guibg=#` or `guibg=#xzdz` but cursor support has always been somewhat a problem for (neo)vim so I trust user to google the solution (like not setting a highlight).

teto commented:
While working on sthg else, I noticed there was a name_to_color function and that in syntax.c hl_group had some robust RgbValue sg_rgb_bg so I managed to use that. It was not possible to use :
```
// highlight groups for 'highlight' option
static garray_T highlight_ga = GA_EMPTY_INIT_VALUE;
```
directly but apparently one should use some kind of cache `static garray_T attr_table = GA_EMPTY_INIT_VALUE;` ? I am not sure how both arrays relate to eachother but I've added some doxygen to help choose the correct parameters anyway (via adding groups).

Advantage is that now the code can even accept color names like "green" and is robust. If I update the highlight group, sometimes the change is taken into account immediately, sometimes, I have to change modes for the cursor color to be updated. I haven't looked into it though as I suspect it's due to this attr_table/highlight_ga dichotomy.
I am not sure about why tests fail (seem stalled), they succeed locally.

zhou13 commented:
For cursor color, you can use `"\x1b]112\x07"` to reset it back to default (reverse fg/bg) in vte/xterm.

It still lacks documents so it is not clear to me how to use  `guicursor`, but it seems that it still hard codes cursor shape for differents mode.  Is it possible to use a vim plugin to swtich the cursor rather than hardcode it?

zhou13 commented on src/nvim/tui/tui.c:
```diff
@@ -463,33 +522,98 @@ static void tui_mouse_off(UI *ui)
+      case SHAPE_HOR:   shape = 3; break;
+      default: WLOG("Unknown shape value %d", shape); break;
+    }
+    printf(TMUX_WRAP("\x1b]50;CursorShape=%d;BlinkingCursorEnabled=%d\x07"),
```
why printf?

justinmk replied:
Should be out () I think

teto replied:
fixed

teto committed:
Taking into account reviews

teto commented:
lint failure is voluntary, I can change it if needed. The other travis failure is due to network connectivy problems (I don't understand the QuickBuild failure)
I updated the guicursor help and replaced the printf with unibi_format (using `out` requires to register 2 unibilium strings, w/o tmux).

> For cursor color, you can use "\x1b]112\x07" to reset it back to default (reverse fg/bg) in vte/xterm.

do you suggest to use it when there is no highlight set in guicursor ?

teto commented:
Thanks for rerunning the travis one, that fixed it. As for quickbuild, seems like it times out at "publish & co": http://neovim-qb.szakmeister.net/build/11755/step_status. I wonder if it's linked to the "ci" folder renaming.

zhou13 commented:
@teto With no highlight setting, I think we shouldn't touch user's cursor color because some crazy terminals such as `xfce4-terminal` uses a colored cursor by default.  However, the user should be able to  set
````
:highlight Cursor cterm=reverse
````
to trigger it.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -2802,7 +2802,8 @@ A jump table for the options with a short description can be found at |Q_op|.
 	horizontal cursor.
-	For a console the 't_SI' and 't_EI' escape sequences are used.
+	For a console, shape is taken into account and color as well if
+	'termguicolors' is set. 't_SI' and 't_EI' are deprecated in neovim.
```
Why is it based on 'termguicolors'? It should set cursor settings if the user set 'guicursor'. Users that don't want that just must not do that.

teto replied:
Right, I could not find a CSI to change cursor color other than with an hex color.

zhou13 commented:
````
attr-list is a comma separated list (without spaces) of the
	following items (in any order):
		bold
		underline
		undercurl	a curly underline
		reverse
		inverse		same as reverse
		italic
		standout
		NONE		no attributes used (used to reset it)
````

I think what I said is wrong.  Therefore, reverse/inverse should generate a complementary color of current bg/fg (that is what I see in nvim-qt).  None should reset the cursor color.

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -463,33 +522,102 @@ static void tui_mouse_off(UI *ui)
+    switch (shape) {
+      case SHAPE_BLOCK: shape = 0; break;
+      case SHAPE_VER:   shape = 1; break;
+      case SHAPE_HOR:   shape = 3; break;
```
for konsole this was 2 before. Is this intentional?

teto replied:
good catch my fault sorry

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -463,33 +522,102 @@ static void tui_mouse_off(UI *ui)
+
+/// Prevents useless output
+/// @param mode editor mode
 static void tui_mode_change(UI *ui, int mode)
```
the main description is odd

justinmk commented:
@teto Nice work, this looks great. But why are there no tests?

justinmk commented:
@teto Just FYI, I'm working on some test coverage for this.

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -7,40 +7,84 @@
+  { "vdrag", 0, 0, 0,   0L,   0L,   0L, 0, 0, "vd", SHAPE_MOUSE },
+  { "more", 0, 0, 0,   0L,   0L,   0L, 0, 0, "m", SHAPE_MOUSE },
+  { "more_lastline", 0, 0, 0,   0L,   0L,   0L, 0, 0, "ml", SHAPE_MOUSE },
+  { "match_paren", 0, 0, 0, 100L, 100L, 100L, 0, 0, "sm", SHAPE_CURSOR },
```
'showmatch' is different than 'matchparen'. Am I missing something?

teto replied:
You must be right I just joined the comment as a description `#define SHAPE_IDX_SM    16      /* showing matching paren */. I was not aware of the difference

teto commented:
I don't mind writing tests but in this case I was not sure how to go about it so I am glad you take over xD

justinmk commented:
Merged. Thanks @teto !
